File: DotNet/Messaging/MessageFetchers/MessageFetcherBase.cs
using System.Collections.Generic;
using System.Linq;
using Bollywell.Hydra.Messaging.MessageIds;
using Bollywell.Hydra.Messaging.Storage;
using LoveSeat;
using LoveSeat.Interfaces;
using Newtonsoft.Json.Linq;

namespace Bollywell.Hydra.Messaging.MessageFetchers
{
    public abstract class MessageFetcherBase<TMessage> : IMessageFetcher<TMessage> where TMessage : TransportMessage
    {
        protected abstract string ViewName { get; }

        protected abstract IKeyOptions MessageKey(IMessageId id);
        protected abstract IKeyOptions EndKey();

        #region Implementation of IMessageFetcher<TMessage>

        public IEnumerable<TMessage> MessagesAfterIdUpToSeq(IStore store, IMessageId startId, long lastSeq)
        {
            return AllMessagesFrom(store, startId).Where(row => (long) row["value"] <= lastSeq).Select(TransportMessage.Hydrate<TMessage>)
                    .SkipWhile(message => message.MessageId.CompareTo(startId) <= 0);
        }

        public IEnumerable<TMessage> MessagesInSet(IStore store, IEnumerable<IMessageId> messageIds)
        {
            var options = new ViewOptions { IncludeDocs = true, Keys = messageIds.Select(MessageKey) };
            return store.GetDocs(ViewName, options).Select(TransportMessage.Hydrate<TMessage>);
        }


        #endregion

        private IEnumerable<JToken> AllMessagesFrom(IStore store, IMessageId fromId)
        {
            // CouchDb startkey is inclusive, so this returns messages including fromId
            var options = new ViewOptions { IncludeDocs = true, StartKey = MessageKey(fromId), EndKey = EndKey() };
            return store.GetDocs(ViewName, options);
        }

    }
}

File: Java/hydra/src/main/java/uk/co/shastra/hydra/messaging/messagefetchers/MessageFetcherBase.java
package uk.co.shastra.hydra.messaging.messagefetchers;

import java.util.ArrayList;

import org.ektorp.ComplexKey;
import org.ektorp.ViewQuery;

import com.fasterxml.jackson.databind.JsonNode;

import uk.co.shastra.hydra.messaging.TransportMessage;
import uk.co.shastra.hydra.messaging.messageids.MessageId;
import uk.co.shastra.hydra.messaging.storage.Store;

public abstract class MessageFetcherBase<TMessage extends TransportMessage> implements MessageFetcher<TMessage> {

    private Class<TMessage> messageType;

	protected abstract String getViewName();
    protected abstract ComplexKey getMessageKey(MessageId id);
    protected abstract ComplexKey getEndKey();
    
    // TODO Constructors for JavaType etc
    protected MessageFetcherBase(Class<TMessage> messageType) {
    	this.messageType = messageType;
    }
    
	@Override
	public Iterable<TMessage> messagesAfterIdUpToSeq(Store store, MessageId startId, long lastSeq) {
		ArrayList<TMessage> res = new ArrayList<TMessage>();
		for (JsonNode json : AllMessagesFrom(store, startId)) {
			if (json.get("value").longValue() > lastSeq) continue;
			TMessage message = TransportMessage.hydrate(json, messageType);
			if (message.getMessageId().compareTo(startId) > 0) res.add(message);
		}
		return res;
	}

	@Override
	public Iterable<TMessage> messagesInSet(Store store, Iterable<MessageId> messageIds) {
		ArrayList<ComplexKey> keys = new ArrayList<ComplexKey>();
		for (MessageId messageId : messageIds) {
			keys.add(getMessageKey(messageId));
		}
    	ViewQuery options = new ViewQuery().includeDocs(true).keys(keys);
    	
    	Iterable<JsonNode> docs = store.getDocs(getViewName(), options);
    	ArrayList<TMessage> res = new ArrayList<TMessage>();
    	for (JsonNode doc : docs) {
    		res.add(TransportMessage.hydrate(doc, messageType));
    	}
    	
        return res;
	}

    private Iterable<JsonNode> AllMessagesFrom(Store store, MessageId fromId)
    {
        // CouchDb startkey is inclusive, so this returns messages including fromId
    	ViewQuery options = new ViewQuery().includeDocs(true).startKey(getMessageKey(fromId)).endKey(getEndKey());
        return store.getDocs(getViewName(), options);
    }
}
